File: platforms/android/ime/src/main/java/app/funput/funput/ime/suggestions/PersonalSuggestionCasing.kt
package app.funput.funput.ime.suggestions

import java.util.Locale

internal object PersonalSuggestionCasing {
    private val Vietnamese = Locale.forLanguageTag("vi-VN")

    /**
     * A prediction has no prefix to take its shape from, so [capitalized] — the
     * keyboard's own Shift state — is the only thing left to ask.
     */
    fun apply(candidate: String, prefix: String, capitalized: Boolean = false): String = when {
        prefix.isEmpty() -> if (capitalized) candidate.titlecased() else candidate
        prefix.hasLetters() && prefix == prefix.uppercase(Vietnamese) -> candidate.uppercase(Vietnamese)
        prefix.firstOrNull()?.isUpperCase() == true -> candidate.titlecased()
        else -> candidate.lowercase(Vietnamese)
    }

    private fun String.titlecased() = replaceFirstChar { char -> char.titlecase(Vietnamese) }

    private fun String.hasLetters() = any(Char::isLetter)
}

File: platforms/ios/Packages/FunputKit/Sources/PersonalSuggestions/Service/PersonalSuggestionCasing.swift
import Foundation

enum PersonalSuggestionCasing {
    private static let vietnamese = Locale(identifier: "vi_VN")

    /// A prediction has no prefix to take its shape from, so `capitalized` — the
    /// keyboard's own Shift state — is the only thing left to ask.
    static func apply(
        _ candidate: String,
        matching prefix: String,
        capitalized: Bool = false
    ) -> String {
        let letters = prefix.filter(\.isLetter)
        guard !letters.isEmpty else {
            guard capitalized else { return candidate }
            return candidate.prefix(1).uppercased(with: vietnamese) + candidate.dropFirst()
        }
        if letters == letters.uppercased(with: vietnamese) {
            return candidate.uppercased(with: vietnamese)
        }
        guard let first = prefix.first,
              String(first) == String(first).uppercased(with: vietnamese)
        else { return candidate }
        return candidate.prefix(1).uppercased(with: vietnamese)
            + candidate.dropFirst()
    }
}
